File: 백준/Silver/2667. 단지번호붙이기/단지번호붙이기.py

def dfs(x,y):
    if x >= T or x <= -1 or y >= T or y <= -1:
        return False
    if graph[x][y] == 1:
        global result
        result += 1
        graph[x][y] = 0
        dfs(x+1,y)
        dfs(x-1,y)
        dfs(x,y+1)
        dfs(x,y-1)
        return True
    return False

T = int(input())
graph = []
num = []

for _ in range(T):
    graph.append(list(map(int,input()))) #2차원 배열 생성
result = 0
counter = 0
for i in range(T):
    for j in range(T):
        if dfs(i,j) == True:
            num.append(result)
            counter += 1
            result = 0
num.sort()
print(counter)
print(*num, sep='\n')

File: 백준/Silver/2667. 단지번호붙이기/단지번호붙이기.js
const readline = require("readline");

const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout,
});

let inputLine = 0;
let N = 0;
let arr = [];
let answer = [];
let visited;
let dx = [-1, 1, 0, 0];
let dy = [0, 0, -1, 1];
let counter = 0;
rl.on("line", (line) => {
  if (inputLine == 0) {
    N = parseInt(line);
    visited = Array.from({ length: N }, () => Array(N).fill(0));
  }
  if (inputLine > 0 && inputLine <= N) {
    arr.push(line.split("").map((el) => parseInt(el)));
  }
  if (inputLine == N) {
    solution();
    process.exit();
  }
  inputLine++;
});

function solution() {
  for (let i = 0; i < arr.length; i++) {
    for (let j = 0; j < arr.length; j++) {
      if (!visited[i][j] && arr[i][j]) {
        dfs(i, j);
        answer.push(counter);
        counter = 0; //counter 초기화
      }
    }
  }
  console.log(answer.length);
  answer.sort((a, b) => a - b);
  console.log(answer.join("\n"));
}

function dfs(x, y) {
  visited[x][y] = true;
  arr[x][y] = 0;
  counter += 1;
  for (let i = 0; i < 4; i++) {
    let nx = x + dx[i];
    let ny = y + dy[i];
    if (nx < 0 || ny < 0 || nx >= N || ny >= N) {
      continue;
    } else if (!visited[nx][ny] && arr[nx][ny]) {
      dfs(nx, ny);
    }
  }
}
